src/mon/command_multiplexer.h

```diff
@@ -0,0 +1,312 @@
+#include <memory>
+#include <algorithm>
+#include <exception>
+#include <string_view>
+
+#include <experimental/iterator>
+
+/*
+This is a mini-framework for mapping commands and their parameters from the command-line (or wherever) to functions.
+
+For example, if you are at the command-line and run "ceph mon add", this command will be forwarded to this framework.
+
+From a high level, this library provides a container, called a command_registry, that holds bindings from strings
+to functions, and manages the details of memory management, standardizing the details of whether or not a command
+was handled, etc.. 
+
+While much of the framework isn't directly specific to mon, there are two major ways involving specific code: we
+have two kinds of operations, preprocess operations and prepare operations. 
+
+Installing a command_registry into a subsystem requires patching the preprocess and prepare pipelines. Additionally,
+the environment a user's function (ie. a command) runs in is centralized into a single state structure.
+
+Implementation note:
+	- rather than having a vector backing everything-- requiring expensive lookups for any search operation because
+    keys in each object must be examined-- it would certainly be possible to back this with a boost::multiindex, or
+    map of shared_ptr<>. Left as a future refinement. If this is done, I suggest making that portion of the implementation
+    a type in common that more closely models an STL associative array, and building this around that implementation.
+*/
+
+namespace ceph {
+
+enum class operation_type : int { preprocess, prepare };
+
+template <typename CommandState, typename CommandResult>
+struct command;
+
+template <typename CommandType>
+struct command_registry
+{
+ using command_type = CommandType;
+
+ // The type returned by an operation:
+ using status_type  = std::optional<typename command_type::result_type>;
+
+ private:
+ command_registry(const command_registry&)              = delete;
+ command_registry& operator=(const command_registry&)   = delete;
+
+ public:
+ const std::string name;      // eg. the name of the registry 
+
+ std::vector<std::unique_ptr<command_type>> commands;
+
+ public:
+ command_registry(std::string_view name_)
+  : name(name_)
+ {}
+
+ virtual ~command_registry() = default;
+
+ public:
+ auto find_cmd_for_tag(std::string_view candidate_tag) const
+ {
+    using std::end;
+    using std::begin;
+
+    auto tag_match = [candidate_tag](const auto& command) {
+        auto tags = command->tags();
+        return end(tags) != std::find(begin(tags), end(tags), candidate_tag);
+    };
+
+    return std::find_if(begin(commands), end(commands), tag_match); 
+ }
+
+ // True if a "tag" is found:
+ bool in_registry(std::string_view tag) const
+ {
+    return end(commands) != find_cmd_for_tag(tag);
+ }
+
+ // True if any tag in the "tags" sequence is found:
+ template <typename K, typename A, template <typename, typename> typename SeqT = std::vector>
```

Comment:
Good idea, thanks! (It will be /extra/ nice to be able to use Concepts for exactly this sort of thing.)

Reply:
Mini-update: I wound up opting for the comparatively straight forward use of a begin, end iterator pair-- since these are working through member functions rather than class types, the SFINAE gets inordinately hairy as both too much and too little matching happens: resolving the overload between containers and string_view gets "fun" and I realized it isn't really what I needed anyway!